app/Main.hs

```diff
@@ -81,24 +87,26 @@ main = do
                                       `catch`
                                       (\e -> do putStrLn $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options
                                                 throw (e::SomeException))
-                   _ -> error $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options 
-
-    let fname  = confDatalogFile config
-        (dir, file) = splitFileName fname
-        (basename,_) = splitExtension file
-        --workdir = dir </> basename
+                   _ -> error $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options
     case confAction config of
-         ActionValidate -> readValidate fname
-         ActionCompile -> error "not implemented"
+         ActionValidate -> do { parseValidate config; return () }
+         ActionCompile -> compileProg config
          ActionNone -> error "action not specified"
 
 
-readValidate :: FilePath -> IO DatalogProgram
-readValidate fname = do
-    prog <- parseDatalogFile True fname
---    case validate prog of
---         Left e  -> error $ "Validation error: " ++ e
---         Right _ -> return ()
-    putStrLn "Validation complete"
-    return prog
-
+parseValidate :: Config -> IO DatalogProgram
+parseValidate Config{..} = do
+    d <- parseDatalogFile True confDatalogFile
+    case validate d of
+         Left e   -> errorWithoutStackTrace $ "error: " ++ e
+         Right d' -> return d'
+
+compileProg :: Config -> IO ()
+compileProg conf@Config{..} = do
+    let specname = takeBaseName confDatalogFile
+    prog <- parseValidate conf
```

Comment:
will this handle errors properly?

Reply:
You mean readFile? It will fail with a meaningful error message. And so should parseValidate